src/core/rate-limiter.js

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2023 RethinkDNS and its authors.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+const RATE_LIMIT_CAPACITY = 300; // max amount of requests at once. Only used if over 30 requests a sec
+const RATE_LIMIT_FILL_RATE = 30; // max amount of requests that can be processed a sec before going to RATE_LIMIT_CAPACITY
+
+class RateLimiter {
+  constructor() {
+    this.bucket = {
+      capacity: RATE_LIMIT_CAPACITY,
+      tokens: RATE_LIMIT_CAPACITY,
+    };
+    setInterval(() => {
```

Comment:
RateLimiters using timers are a very inefficient way to implement it. Besides, I believe, rate limiting is best done out of process, like in HA Proxy fronting Node, though I'm undecided yet.